### ISSUE TITLE
Renaming and redefining ceterms:SelectionComponent

### ISSUE THREAD

**Issue by stuartasutton** [CredentialEngine/Schema-Development#804]
Subsequent to our discussion of the distinction between [ceterms:ComponentCondition](https://purl.org/ctdl/terms/ComponentCondition) and [ceterms:SelectionComponent](https://purl.org/ctdl/terms/SelectionComponent), my conclusion is that the unstable [ceterms:SelectionComponent](https://purl.org/ctdl/terms/SelectionComponent) needs renaming and redefinition to make its use clear.

>**Current Name:** [ceterms:SelectionComponent](https://purl.org/ctdl/terms/SelectionComponent)
**Current Definition:** "Component which provides a condition which references an array of other components from which components must be selected."

Suggested changes:
>**Proposed Name Change:** `AggregateComponent`
>**Proposed Definition Change:** "Resource that identifies a set of candidate components from which a specified number must be selected to satisfy the component."

**Comment by siuc-nate** [comment]
As we discussed today and to expand on my point a little:
The entire reason we created Selection Component was to compensate for not being able (for better or worse) to construct a hierarchy of Component Conditions.  The [Badgr use case](https://credentialfinder.org/pathway/3/French_B_A__Marketable_Skills) that led us to creating Selection Component has places where the user must make a choice within a choice, and structurally we couldn't handle that since Component Condition only went one layer deep.  We also wanted to mirror the structure of the source data closely, and having Selection Component as a standalone node allowed us to do that.

However, an alternative to using Selection Component would be to enable Component Condition to have a hierarchical structure (by adding `ceterms:ComponentCondition` to the domain of `ceterms:hasCondition`).  Doing so would (I believe) negate the need for Selection Component (and, by extension, negate the need for AggregateComponent).

For example:
```
{
  "@type": "ceterms:CredentialComponent",
  "ceterms:name": { "en": "Interpersonal Communication" },
  "ceterms:hasCondition": [
    {
      "@type": "ceterms:ComponentCondition",
      "ceterms:requiredNumber": 4,
      "ceterms:description": { "en": "Requires 4 of 5" },
      "ceterms:hasCondition": [
        {
          "@type": "ceterms:ComponentCondition",
          "ceterms:requiredNumber": 2,
          "ceterms:name": { "en": "French 2040/2050" },
          "ceterms:description": { "en": "Requires 4 of 5" },
          "ceterms:targetComponent": 
            "https://uri-for-French-2040-Guided-Writing",
            "https://uri-for-French-2040-Oral-Participation",
            "https://uri-for-French-2050-In-Class-Oral-Presentation",
            "https://uri-for-French-2050-In-Class-Oral-Participation"
          ]
        },
        {
          "@type": "ceterms:ComponentCondition",
          "ceterms:requiredNumber": 1,
          "ceterms:name": { "en": "French 3040/3045" },
          "ceterms:description": { "en": "Requires 1 of 2" },
          "ceterms:targetComponent": 
            "https://uri-for-French-3040",
            "https://uri-for-French-3045"
          ]
        },
        {
          "@type": "ceterms:ComponentCondition",
          "ceterms:requiredNumber": 1,
          "ceterms:name": { "en": "French 3050/3055" },
          "ceterms:description": { "en": "Requires 1 of 2" },
          "ceterms:targetComponent": 
            "https://uri-for-French-3050",
            "https://uri-for-French-3055"
          ]
        },
        {
          "@type": "ceterms:ComponentCondition",
          "ceterms:requiredNumber": 1,
          "ceterms:name": { "en": "French 3060/3065" },
          "ceterms:description": { "en": "Requires 1 of 2" },
          "ceterms:targetComponent": 
            "https://uri-for-French-3060",
            "https://uri-for-French-3065"
          ]
        },
        {
          "@type": "ceterms:ComponentCondition",
          "ceterms:requiredNumber": 1,
          "ceterms:name": { "en": "French 3070/3075/3095" },
          "ceterms:description": { "en": "Requires 1 of 3" },
          "ceterms:targetComponent": 
            "https://uri-for-French-3070",
            "https://uri-for-French-3075",
            "https://uri-for-French-3095",
          ]
        }
      ]
    }
  ]
}
```

I want to try this with the TSAP nursing pathway as well before I commit to it as an alternative, but I think it may be viable.

**Comment by stuartasutton** [comment]
@siuc-nate, I updated your JSON above...missing opening bracket ("[") in a number of places.

If this works out with TSAP, it's a more elegant solution than the new type of Pathway Component under discussion, no matter how named. I've no objections to the use of [ceterms:hasCondition](https://credreg.net/ctdl/terms#hasCondition).

**Comment by siuc-nate** [comment]
Good catch - a by-product of copying/pasting, I think.

As for the TSAP pathway, I think it can be done much more simply than I had originally modeled it, not using SelectionComponents at all:
https://app.diagrams.net/#G1OdF97mP8euAqBSCiJIaQJ7Q5Q7OYxjpv ("Revision" tab)

**Comment by jeannekitchens** [comment]
@siuc-nate provide an example using this published badgr pathway https://credentialfinder.org/pathway/3/French_B_A__Marketable_Skills

**Comment by stuartasutton** [comment]
Nate, here is a revised model that handles nesting of conditions as you note above. However, instead of reusing the [hasCondition](https://purl.org/ctdl/terms/hasCondition) property, I suggest that we create a new property aligned to [dcterms:conformsTo](http://purl.org/dc/terms/conformsTo), but rather than referencing a "standards" it references a conformant rule (perhaps `ceterm:conformantRule`).

![Alternative Rule Sets-Using Rules](https://user-images.githubusercontent.com/2939046/147771022-ab88158c-89a4-4397-aa55-cd4a724b8494.png)

**Comment by siuc-nate** [comment]
The only rule in play in the example is "requiredNumber", it's just that it applies at two layers (hence the reuse of hasCondition and the nesting that goes along with it).  Are you saying we would need data like what's in your diagram to handle that example instead? It seems more complex than it needs to be.

**Comment by stuartasutton** [comment]
Nate, then you are misinterpreting the diagram or the diagram is not clear. The "required number" is not the only "rule in play". The diagram reads, "Select 3 out of these 6 course components (rule 1) **WHERE** those components have an academic level equal to "Upper Division" (rule 2) AND credit hours of "2" or more credits (rule 3)".

**Comment by siuc-nate** [comment]
Sorry, I didn't notice that there was a different example encoded into the diagram.  By "example", I meant the French BA pathway [that was solved](https://github.com/CredentialEngine/Schema-Development/issues/804#issuecomment-970613525) by having nested conditions.  How would the structure you're proposing handle that example?

Also, is the example in your diagram based on real data? I'd like to see it in the context of a pathway to get a better handle on it.

**Comment by siuc-nate** [comment]
One more thing - did you intend to encode the integer `2` as a skos:Concept in your example? It seems like we wouldn't want a skos:Concept for each of a bunch of numbers (and likely decimal'd numbers between integers).

**Comment by stuartasutton** [comment]
I though the answer to how it would comport (https://github.com/CredentialEngine/Schema-Development/issues/804#issuecomment-1004220757) would be answered by your working through several examples. I think that my modeling comports with your example but doesn't repeat the use of `hasCondiiton` and instead uses `conformatRule`.

**Comment by stuartasutton** [comment]
Good catch/comment, Nate (https://github.com/CredentialEngine/Schema-Development/issues/804#issuecomment-1004222490) regarding numbers as opposed to skos:Concept.

**Comment by siuc-nate** [comment]
Responses to this presentation:
https://docs.google.com/presentation/d/1FQAG6gXW4IFr1R4p6_FmJMXqUYbcL2MgnXxR1gQYkl4/edit#slide=id.p

Regarding this slide:
https://docs.google.com/presentation/d/1FQAG6gXW4IFr1R4p6_FmJMXqUYbcL2MgnXxR1gQYkl4/edit#slide=id.g1150f985a98_0_19
>...the Constraint entity defines precise limitations on the attributes of the components in the array that each must satisfy.

The purpose of the component condition (and, by extension, the constraints attached to it) is to explain what the person using the pathway must satisfy, not what the components themselves must satisfy. Components that aren't valid options wouldn't be pointed to by the component condition in the first place.

---

Regarding this diagram:
https://docs.google.com/presentation/d/1FQAG6gXW4IFr1R4p6_FmJMXqUYbcL2MgnXxR1gQYkl4/edit#slide=id.g1150f985a98_0_72
![image](https://user-images.githubusercontent.com/21346519/154095476-c78fffc3-b9c5-497b-ad12-2d6199751b0b.png)
- If requiredNumber = 1, and the component condition only points to the assessment component, then no constraint is needed, since you are required to complete the assessment regardless of its nature.

---

Regarding this diagram:
https://docs.google.com/presentation/d/1FQAG6gXW4IFr1R4p6_FmJMXqUYbcL2MgnXxR1gQYkl4/edit#slide=id.g1150f985a98_0_107
![image](https://user-images.githubusercontent.com/21346519/154095030-5cb7f842-1cce-45ef-80ea-93f3a1d4b975.png)
- The first constraint is unnecessary because the component condition has requiredNumber = 1 and only points to one target component.
- The other two sets of conditions could be simplified: The user is required to complete at least two of (course B, C, or D) regardless, and may optionally also complete course A.
- The constraints are only necessary if it is possible to take two of (course B, C, or D) but still not have earned at least 2 credits in doing so, which doesn't seem logical.
- It would help if the diagram showed how much credit each of the course components was worth.

**Comment by stuartasutton** [comment]
This issue is resolved in the changes summary of #807–deprecate [ceterms:SelectionComponent](https://purl.org/ctdl/terms/SelectionComponent).

**Comment by stuartasutton** [comment]
This issue is resolved in the changes summary of #807–deprecate [ceterms:SelectionComponent](https://purl.org/ctdl/terms/SelectionComponent).

**Comment by stuartasutton** [comment]
This issue is resolved in the changes summary of https://github.com/CredentialEngine/Schema-Development/issues/807–deprecate [ceterms:SelectionComponent](https://purl.org/ctdl/terms/SelectionComponent).
